### ISSUE TITLE
clustersecretstore getting 403 from external vault but vault cli is not

### ISSUE THREAD

**Issue by the-scott-hand** [external-secrets/external-secrets#2691]
looking for some help, not sure where to look next for debugging this issue i am having. any insight yall can provide would be much appreciated!

`image: ghcr.io/external-secrets/external-secrets:v0.8.3`

I have two kubernetes clusters, one running hashicorp vault(cluster A) and the other running with external secrets(cluster B). Unfortunately my clustersecretstore in cluster-B gives the following issue:
```txt
Warning InvalidProviderConfig cluster-secret-store unable to log in to auth method: unable to log in with Kubernetes auth: Error making API request.
URL: PUT https://<url>/v1/auth/cluster-name/kubernetes/login
Code: 403. Errors:
* permission denied

Status:
  Message: unable to create client
```
 I am using the kubernetes auth method for a clustersecretstore resource with a spec that looks like this:

```yaml
spec:
  provider:
    vault:
      auth:
        kubernetes:
          mountPath: cluster-name/kubernetes
          role: reader
          secretRef:
            key: token
            name: vault-token
            namespace: project-name
          serviceAccountRef:
            name: vault
            namespace: project-name
      caProvider:
        key: ca.crt
        name: vault-ca
        namespace: project-name
        type: Secret
      path: cluster-name/stuff
      server: https://<url>
      version: v2
```
I have confirmed I can connect to the vault myself by creating the following test pod in cluster-B
```yaml
# test pod.yaml
apiVersion: v1
kind: Pod
metadata:
  name: vault-test
spec:
  containers:
  - args:
    - -c
    - sleep 100000
    command:
    - /bin/sh
    image: vault:1.12.3
    name: vault
```
and exec-ing in to run the following command in this pod from cluster-B:
```bash
kubectl exec -it vault-test -- sh
vault write -address=https://<url> -tls-skip-verify auth/cluster-name/kubernetes/login role=reader jwt=<jwt-from-vault-token-secret>
# SUCCESS!
# note, i have also done this with -ca-cert=ca.crt using my ca with the same success
```
This succeeds. But the clustersecretstore cannot seem to authenticate using the same url and the same jwt. I have turned on debug log level in the external-secrets pod and it says this before stacktrace. 
```txt
Retrieved new token using Kubernetes auth.
unable to validate store, could not get provider client... etc
``` 

Any ideas why a test pod in cluster-B with the vault cli can authenticate, but the clustersecretstore in cluster-B cannot? I have also tried giving external-secrets cluster-admin clusterrole but without success either. 

Thanks again for your help!

**Comment by moolen** [comment]
You specify both kubernetes.secretRef and kubernetes.serviceAccountRef, is that intentional?

IIRC the service account takes precedence. From your test pod, it looks like you've used the vault token from the secret?! `jwt=<jwt-from-vault-token-secret>`
```yaml
spec:
  provider:
    vault:
      auth:
        kubernetes:
          mountPath: cluster-name/kubernetes
          role: reader
          secretRef:
            key: token
            name: vault-token
            namespace: project-name
          serviceAccountRef:
            name: vault
            namespace: project-name
```

You get a 403, that indicates that you are authenticated, but there's no policy that allows that action. Do you have access to the vault (debug) logs?

**Comment by moolen** [comment]
Looking at the code: the kubernetes `Login()` call does a [WriteWithContext](https://github.com/hashicorp/vault/blob/main/api/auth/kubernetes/kubernetes.go#L73-L84) which does a [MethodPUT](https://github.com/hashicorp/vault/blob/main/api/logical.go#L207)! The docs say that [only POST is supported](https://developer.hashicorp.com/vault/api-docs/auth/kubernetes#login). But also: `Vault treats the HTTP POST and PUT verbs as equivalent` [from here](https://developer.hashicorp.com/vault/docs/concepts/policies).

I don't think POST is needed here, i'd guess it is a authn or authz issue: either a misconfigured vault policy or misconfigured kubernetes integration in vault.

**Comment by the-scott-hand** [comment]
Good catch on us using two methods of auth! We ended up for now just using the vault token from the secret since we accidentally configured 2 methods of auth. This is working for us now.

Our preference was the serviceaccount method, and after some debugging we suspect our particular set of systems may have an issue with clock skew. By investigating the vault audit logs we were able to find the token generated by the service account. Running the commands manually with this token also resulted in success, but failure when run automatically. The running theory was something along the lines of clock skew causing the validFrom date to be in the future, meaning the api would not yet accept this token as valid but would in a few seconds. This explained why using it manually later worked, but the automation failed because it was too quick given our clock skew. 

When it fails a new token seems to be generated rather than the previous one being retried. If the previous one was retried a couple times that in theory would give some extra allowance for this condition. At some point I found some caching property I thought might help, but at that point we had something working and moved on. Reality is we should probably just try fixing clock skew issue, but in the meantime the secretRef is working until that becomes a priority.

Thanks for the help!
